File: baekjoon/src/dfsAndBfs/No2606.java
package dfsAndBfs;

import java.io.*;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

/**
 * https://www.acmicpc.net/problem/2606
 */
public class No2606 {
    public static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    public static final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
    public static int nodes;
    public static int connections;
    public static final int startAt = 1;
    public static boolean[] isVisited;
    public static int[][] graph;

    public static void setGraph() throws IOException {
        StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
        nodes = Integer.parseInt(tokenizer.nextToken());
        tokenizer = new StringTokenizer(reader.readLine());
        connections = Integer.parseInt(tokenizer.nextToken());
        isVisited = new boolean[nodes + 1];
        graph = new int[nodes + 1][nodes + 1];

        for (int i = 0; i < connections; i++) {
            tokenizer = new StringTokenizer(reader.readLine());
            int source = Integer.parseInt(tokenizer.nextToken());
            int destination = Integer.parseInt(tokenizer.nextToken());
            graph[source][destination] = graph[destination][source] = 1;
        }
    }

    public static int countWormVirus(int startAt) {
        int wormVirus = 0;
        Queue<Integer> queue = new LinkedList<>();
        queue.add(startAt);
        isVisited[startAt] = true;

        while (!queue.isEmpty()) {
            int number = queue.poll();
            for (int i = 1; i <= nodes; i++) {
                if (graph[number][i] == 1 && !isVisited[i]) {
                    queue.add(i);
                    isVisited[i] = true;
                    wormVirus++;
                }
            }
        }
        return wormVirus;
    }

    public static void main(String[] args) throws IOException {
        setGraph();
        writer.write(String.valueOf(countWormVirus(startAt)));
        writer.flush();
    }
}

File: baekjoon-classification/src/dfs/No2606.kt
package dfs

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.StringTokenizer

/**
 * https://www.acmicpc.net/problem/2606
 */
private var count = 0

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val nodes = readLine().toInt()
    val edges = readLine().toInt()
    val graph = Array(nodes + 1) { IntArray(nodes + 1) }
    val isVisited = BooleanArray(nodes + 1)

    var tokenizer: StringTokenizer
    repeat(edges) {
        tokenizer = StringTokenizer(readLine())
        val row = tokenizer.nextToken().toInt()
        val col = tokenizer.nextToken().toInt()
        graph[row][col] = 1
        graph[col][row] = 1
    }

    dfs(graph, isVisited, 1)

    with(BufferedWriter(OutputStreamWriter(System.out))) {
        write("${count - 1}")
        flush()
    }
}

private fun dfs(graph: Array<IntArray>, isVisited: BooleanArray, index: Int) {
    isVisited[index] = true
    count++

    for (i in 1 until graph.size) {
        if (!isVisited[i] && graph[index][i] == 1)
            dfs(graph, isVisited, i)
    }
}